File: leetcode/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
public:
    vector<string>ans;
    void dfs(string tmp, int left, int right){
        if(left == 0 && right == 0){
            ans.push_back(tmp);
            return;
        }
        if(left>right){
            return;
        }
        if(left>0){
            dfs(tmp+'(',left-1,right);
        }
        if(right>0){
            dfs(tmp+')',left, right-1);
        }
    }
    vector<string> generateParenthesis(int n) {
        if(n==0)return ans;
        dfs("", n ,n);
        return ans;
    }
};
// @lc code=end

File: leetcode/22. 括号生成.py
# 数字 n 代表生成括号的对数，请你设计一个函数，用于能够生成所有可能的并且 有效的 括号组合。

# 题解： 先加入左括号，再加入右括号，如果左括号的数量大于右括号，则加入左括号，否则加入右括号。

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:

        res=[]
        
        def dfs(s, left, right):
            if left==0 and right==0:
                res.append(s[:])
                return

            if left>0:
                dfs(s+"(",left-1,right)
            if left<right:
                dfs(s+")",left, right-1)
        dfs("",n,n)
        return res

        